### PR TITLE
US014: Parking Slot Index Action

### PR DESCRIPTION
**Pull request by iamurtazo**

### CODE REVIEW

File: ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        _slotService = _slot;
+        _zoneService = _zone;
+    }
+
+    #region Index
+    // GET: Admin/ParkingSlot
+    public IActionResult Index(int id)
+    {
+        var parkingSlotVMs = _slotService.GetSlots(id)
+                                         .Select(slot => new ListOfParkingSlotsVM(slot));
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs]: So big space for Select

File: ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    public ParkingSlotController(IParkingSlotService _slot, IParkingZoneService _zone)
+    {
+        _slotService = _slot;
+        _zoneService = _zone;
+    }
+
+    #region Index
+    // GET: Admin/ParkingSlot
+    public IActionResult Index(int id)
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs]: It is better if you make id => ParkingZoneId because we will be confused about whether it is Id of Slot or ParkingZone

File: ParkingZoneApp/Areas/Admin/Views/ParkingSlot/Index.cshtml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                @Html.DisplayNameFor(model => model.Number)
+            </th>
+            <th>
+                @Html.DisplayNameFor(model => model.IsAvailableForBooking)
+            </th>
+            <th>
+                @Html.DisplayNameFor(model => model.Category)
+            </th>  
+            <th>
+                Actions
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Areas/Admin/Views/ParkingSlot/Index.cshtml:28]: Why did you add it?

File: ParkingZoneApp/Areas/Admin/Views/ParkingSlot/Index.cshtml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+@{
+    ViewData["Title"] = "Index";
+    Layout = "~/Views/Shared/_Layout.cshtml";
+}
+
+<h1>Index</h1>
+
+<p>
+    <a asp-controller="ParkingSlot" asp-action="Create" class="btn btn-primary">Create New</a>
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Areas/Admin/Views/ParkingSlot/Index.cshtml]: Where is Back button? After entering into Slots you should be able to go back to ParkingZone Index.

File: ParkingZoneApp/Data/ApplicationDbContext.cs [hunk 5]
```diff
@@ -9,14 +9,17 @@ public class ApplicationDbContext : IdentityDbContext
...
-    protected override void OnModelCreating(ModelBuilder builder)
-    {
-        base.OnModelCreating(builder);
-        // Customize the ASP.NET Identity model and override the defaults if needed.
-        // For example, you can rename the ASP.NET Identity table names and more.
-        // Add your customizations after calling base.OnModelCreating(builder);
-    }
+    public DbSet<ParkingSlot> ParkingSlots { get; set; }
+
+    //protected override void OnModelCreating(ModelBuilder modelBuilder)
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Data/ApplicationDbContext.cs]: delete these comments - Uneccessary lines

File: ParkingZoneApp/Services/ParkingSlotService/ParkingSlotService.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        _slotRepository = repository;
+    }
+
+    public IEnumerable<ParkingSlot> GetSlots(int parkingZoneId)
+    {
+        return _slotRepository
+                        .GetAll()
+                        .Where(slot => slot.ParkingZoneId == parkingZoneId);
+    }
+
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Services/ParkingSlotService/ParkingSlotService.cs]: Space

File: ParkingZoneApp/Services/ParkingSlotService/ParkingSlotService.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    private readonly IParkingSlotRepository _slotRepository;
+    public ParkingSlotService(IParkingSlotRepository repository) : base(repository)
+    {
+        _slotRepository = repository;
+    }
+
+    public IEnumerable<ParkingSlot> GetSlots(int parkingZoneId)
+    {
+        return _slotRepository
+                        .GetAll()
```
Comment (SamDotNetDev) [inline ParkingZoneApp/Services/ParkingSlotService/ParkingSlotService.cs]: I think here is big space make it little

**Review by SamDotNetDev — CHANGES_REQUESTED** [review]

File: ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+namespace ParkingSlots.Tests.Controller;
+
+public class ParkingSlotControllerTests
+{
+    private readonly Mock<IParkingSlotService> _slotService;
+    private readonly Mock<IParkingZoneService> _zoneService;
+    private readonly ParkingSlotController _controller;
+    private readonly ParkingZone _parkingZone;
+    private readonly List<ParkingSlot> _parkingSlot;
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs]: Space

File: ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        _parkingSlot = new List<ParkingSlot>
+        {
+            new ()
+            {
+                Id = 1,
+                Number = 2,
+                IsAvailableForBooking = false,
+                ParkingZoneId = 1,
+                Category = 0,
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs]: Space

File: ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            new ()
+            {
+                Id = 2,
+                Number = 3,
+                IsAvailableForBooking = true,
+                ParkingZoneId = 1,
+                Category = SlotCategory.Premium
+            }
+        };
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs]: Space

File: ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    #region Index
+    [Fact]
+    public void GivenParkingZoneId_WhenIndexIsCalled_ThenReturnsParkingSlotsVM()
+    {
+        // Arrange
+        var expectedVMs = new List<ListOfParkingSlotsVM>();
+        expectedVMs.AddRange(ListOfParkingSlotsVM.MapToModel(_parkingSlot));
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs:56]: Space

File: ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        _slotService
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs]: make it in one line

File: ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+            .Setup(x => x.GetSlots(_Id))
+            .Returns(_parkingSlot);
+
+        //Act
+       var index = _controller.Index(_Id);
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Controller/ParkingSlotControllerTests.cs]: bitta orqaga surilib ketibdi

File: ParkingSlots.Tests/Models/CreateViewModelTest.cs [hunk 14]
```diff
@@ -0,0 +1,11 @@
+﻿using ParkingZoneApp.Enums;
+using ParkingZoneApp.ViewModels.ParkingSlotViewModels;
+using System.ComponentModel.DataAnnotations;
+
+namespace ParkingSlots.Tests.Models;
+
+public class CreateViewModelTest
+{
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Models/CreateViewModelTest.cs]: where are create vm tests

File: ParkingSlots.Tests/Services/ParkingSlotServiceTest.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    #endregion
+
+    #region GetSlots
+    [Fact]
+    public void GivenParkingZoneId_WhenGetSlotsIsCalled_ThenReturnsParkingSlots()
+    {
+        //Arrange
+        var expected = new List<ParkingSlot>() { _parkingSlot };
+        _parkingSlotRepository.Setup(x => x.GetAll()).Returns(new List<ParkingSlot>() { _parkingSlot });
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Services/ParkingSlotServiceTest.cs:103]: Space

File: ParkingSlots.Tests/Services/ParkingSlotServiceTest.cs [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+        //Act
+        var result = _service.GetSlots(_id);
+
+        //Assert
+        Assert.IsAssignableFrom<IEnumerable<ParkingSlot>>(result);
+        Assert.Equal(JsonSerializer.Serialize(result), JsonSerializer.Serialize(expected));
+        Assert.NotNull(result);
+        _parkingSlotRepository.Verify(x => x.GetAll(), Times.Once);
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Services/ParkingSlotServiceTest.cs]: Space

File: ParkingSlots.Tests/Services/ParkingSlotServiceTest.cs [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+        //Act
+        var getAll = _service.GetAll();
+
+        //Assert
+        Assert.True(JsonSerializer.Serialize(getAll) == JsonSerializer.Serialize(parkingSlots));
+        Assert.IsAssignableFrom<IEnumerable<ParkingSlot>>(parkingSlots);
+        Assert.NotNull(getAll);
+    }
+    #endregion
+
```
Comment (SamDotNetDev) [inline ParkingSlots.Tests/Services/ParkingSlotServiceTest.cs]: 2 Spaces

**Review by SamDotNetDev — CHANGES_REQUESTED** [review]

File: ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs (hunk 1 above)
Comment (iamurtazo) [reply ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs]: If i put all that in one line then it will be hard to read. And in visual studio it looks pretty ok

File: ParkingZoneApp/Areas/Admin/Views/ParkingSlot/Index.cshtml (hunk 3 above)
Comment (iamurtazo) [reply ParkingZoneApp/Areas/Admin/Views/ParkingSlot/Index.cshtml:28]: Actually Abdulloh bro was ok with this, coz he also added this

File: ParkingSlots.Tests/Models/CreateViewModelTest.cs (hunk 14 above)
Comment (iamurtazo) [reply ParkingSlots.Tests/Models/CreateViewModelTest.cs]: i have not moved to next US which is create action. And adding unit test to create vm will be implemented in the next US. 
this CreateViewModelTest class is just added. If you want I will delete it.

**Review by SamDotNetDev — APPROVED** [review]

File: ParkingZoneApp/Services/ParkingSlotService/IParkingSlotService.cs [hunk 18]
```diff
@@ -0,0 +1,8 @@
+﻿using ParkingZoneApp.Models;
+
+namespace ParkingZoneApp.Services.ParkingSlotService;
+
+public interface IParkingSlotService : IService<ParkingSlot>
+{
+    IEnumerable<ParkingSlot> GetSlots(int parkingZoneId);
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Services/ParkingSlotService/IParkingSlotService.cs]:
```suggestion
    IEnumerable<ParkingSlot> GetSlotsByZoneId(int parkingZoneId);
```

Comment (abdullokhtolibjonov) [reply ParkingZoneApp/Services/ParkingSlotService/IParkingSlotService.cs]: name should by fully understandable, (naming convention)

File: ParkingZoneApp/ViewModels/ParkingSlotViewModels/CreateViewModel.cs [hunk 19]
```diff
@@ -0,0 +1,32 @@
+﻿using ParkingZoneApp.Enums;
+using ParkingZoneApp.Models;
+using System.ComponentModel.DataAnnotations;
+
+namespace ParkingZoneApp.ViewModels.ParkingSlotViewModels;
+
+public class CreateViewModel
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/ViewModels/ParkingSlotViewModels/CreateViewModel.cs]: lets remove this file because in this feature we dont need it

File: ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+                                .GetSlots(parkingZoneId)
+                                .Select(slot => new ListOfParkingSlotsVM(slot));
+
+        return View(parkingSlotVMs);
+    }
+    #endregion
+
+    #region Create
+    #endregion
+
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Areas/Admin/Controllers/ParkingSlotController.cs]: line

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

**Review by abdullokhtolibjonov — APPROVED** [review]

File: ParkingSlots.Tests/Models/ListOfParkingSlotsVMTest.cs [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+        };
+
+    [Theory]
+    [MemberData(nameof(Data))]
+    public void GivenItemsToBeValidated_WhenListOfItemsViewModel_ThenValidationIsPerformed
+        (int parkingZoneId, int number, bool isAvailableForBooking, SlotCategory category, bool expected)
+    {
+        // Arrange
+        var createViewModel = new CreateViewModel
+        {
```
Comment (abdullokhtolibjonov) [inline ParkingSlots.Tests/Models/ListOfParkingSlotsVMTest.cs:29]: tests failed

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

**Review by abdullokhtolibjonov — APPROVED** [review]
